cybench/datasets/torch_dataset.py

```diff
@@ -0,0 +1,207 @@
+            df_all_dates = self._dfs_x[KEY_CROP_SEASON][["cutoff_date"]].copy()
+            df_all_dates["date"] = df_all_dates.apply(
+                lambda r: pd.date_range(
+                    end=r["cutoff_date"],
+                    periods=self._max_season_window_length,
+                    freq="D",
+                ),
+                axis=1,
+            )
+            df_all_dates = df_all_dates.explode("date").drop(columns=["cutoff_date"])
+            df_all_dates.reset_index(inplace=True)
+            df_all_dates.set_index([KEY_LOC, KEY_YEAR, "date"], inplace=True)
+
+            # merge to get all dates in time series data
+            df_ts = df_ts.merge(
+                df_all_dates, how="outer", on=[KEY_LOC, KEY_YEAR, "date"]
+            )
+            del df_all_dates
+
+            # NOTE: interpolate fills data in forward direction.
+            df_ts = df_ts.sort_index().interpolate(method="linear")
+            # fill NAs in the front with 0.0
+            df_ts.fillna(0.0, inplace=True)
+
+            # aggregate
+            df_ts.reset_index(inplace=True)
+            if self._aggregate_time_series_to == "week":
+                df_ts["week"] = df_ts["date"].dt.isocalendar().week
+            else:
+                df_ts["dekad"] = df_ts.apply(
+                    lambda r: dekad_from_date(r["date"]), axis=1
+                )
+
+            ts_aggrs = {k: TIME_SERIES_AGGREGATIONS[k] for k in TIME_SERIES_PREDICTORS}
+            # Primarily to avoid losing the "date" column.
+            ts_aggrs["date"] = "min"
+            df_ts = (
+                df_ts.groupby([KEY_LOC, KEY_YEAR, self._aggregate_time_series_to])
+                .agg(ts_aggrs)
+                .reset_index()
+            )
+            df_ts.drop(columns=[self._aggregate_time_series_to], inplace=True)
+
+            # Add time series to self._dfs_x
+            self._dfs_x["combined_ts"] = df_ts.set_index(
+                [KEY_LOC, KEY_YEAR, "date"]
+            ).sort_index()
+        else:
+            self._dfs_x = dataset._dfs_x
+
+        # Bool value that specifies whether missing data values are allowed
+        # For now always set to False
+        self._allow_incomplete = False
+
+    def __getitem__(self, index) -> dict:
+        """
+        Get a sample from the dataset
+        Data is cast to PyTorch tensors where required
+        :param index: index that is passed to the dataset
+        :return: a dict containing the data sample
+        """
+        sample = super(TorchDataset, self).__getitem__(index)
+        if self._aggregate_time_series_to is not None:
+            if self._aggregate_time_series_to == "week":
+                num_time_steps = int(np.ceil(self._max_season_window_length / 7))
+            else:
+                num_time_steps = int(np.ceil(self._max_season_window_length / 10))
+
+            # make sure time series data has the same number of time steps
```

Comment:
Is this not already enforced during initialization?

Reply:
It's not. We could do, but probably less straightforward.